### PR TITLE
V1.2.1 Introduce JSX

### PR DESCRIPTION
**Pull request by OleksiyRudenko**
Key changes:
- introduce JSX
- `data/openWeatherMapAPI/getIconPropertiesFromCode`: return image properties rather than image tag
- `framework/render.js`: memorize component function and mounting point to enable default re-render
- `framework/element.js`: add JSX parser callbacks

Next step:
- add dummy components to cover more JSX-related use cases

### CODE REVIEW

File: src/components/WeatherResults.js [hunk 1]
```diff
@@ -72,31 +85,31 @@ function WeatherForecast() {
...
     const forecastItems = forecastData.map(forecastDataItem => {
       const preparedForecastDataItem = getPreparedForecastData(forecastDataItem);
       return WeatherForecastItem(preparedForecastDataItem);
     });
-    content += forecastItems.join('');
+    content = [...content, ...forecastItems];
   }
 
-  return content ? `<div>${content}</div>` : '';
+  return content.length ? <div>{content}</div> : '';
```
Comment (nktssh) [inline src/components/WeatherResults.js:112]:
```suggestion
  return content.length ? <div>{content}</div> : null;
```

File: src/components/WeatherResults.js [hunk 2]
```diff
@@ -51,18 +56,26 @@ function WeatherToday() {
...
+        </div>
+        <div>
+          <img {...weatherIcon} /> {main} ({description}). Temperature is {tempInUnits}
+        </div>
+      </>
+    );
   }
 
-  return content ? `<div>${content}</div>` : '';
+  return content ? <div>{content}</div> : '';
```
Comment (nktssh) [inline src/components/WeatherResults.js:72]:
```suggestion
  return content ? <div>{content}</div> : null;
```

File: src/components/WeatherResults.js [hunk 3]
```diff
@@ -23,17 +26,19 @@ function WeatherResults() {
...
+          <WeatherToday />
+          <br />
+          <WeatherForecast />
+        </>
+      );
     }
   }
 
-  return `<p>${content}</p>`;
+  return <p>{content}</p>;
```
Comment (nktssh) [inline src/components/WeatherResults.js:41]: I'd make `content = null` by default

File: src/framework/element.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
+/**
+ * Creates DOM Node. Implements jsx-parser's createElement API
+ * @param tag - HTML tag as string or Component function
+ * @param props - element properties as parsed by jsx-parser
+ * @param children - child elements
+ * @returns {DocumentFragment|Element}
+ */
+export const createElement = (tag, props, ...children) => {
+  if (typeof tag === 'function') return tag({ ...props, children }, children);
```
Comment (nktssh) [inline src/framework/element.js:9]: Don't we require curly braces for `if` statements?

**Review by nktssh — APPROVED** [review]

**Review by VitaliiPluto — APPROVED** [review]

File: src/components/WeatherResults.js (hunk 3 above)
Comment (OleksiyRudenko) [reply src/components/WeatherResults.js:41]: Good point. Also managed to get rid of ternaries

File: src/framework/element.js (hunk 4 above)
Comment (OleksiyRudenko) [reply src/framework/element.js]: Fixed code style.
Relied on linter. Trust but verify

**Review by nktssh — APPROVED** [review]

**Pull request merged**